### PR TITLE
[Auto Suggest] OpenSearch SQL autosuggest with ANTLR

### PR DESCRIPTION
**Pull request by mengweieric**
### Description

OpenSearch SQL auto suggest support for query editor.

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by mengweieric** [comment]
need to unblock @paulstn for DQL, will follow up on increasing test coverage.

**Review by joshuali925 — APPROVED** [review]

File: src/plugins/data/public/antlr/shared/utils.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  connectionsService,
+  dataSourceReuqstHandler: ({
+    dataSourceId,
+    title,
+  }: IDataSourceRequestHandlerParams) => Promise<any>,
+  defaultReuqstHandler: () => any
+) =>
+  connectionsService.getSelectedConnection$().pipe(
+    distinctUntilChanged(),
+    switchMap((connection) => {
```
Comment (RyanL1997) [inline src/plugins/data/public/antlr/shared/utils.ts:24]: Is this the place to get multi datasource id?

File: src/plugins/data/public/antlr/opensearch_sql/grammar/.antlr/OpenSearchSQLLexer.java [hunk 2]
```diff
@@ -0,0 +1,2836 @@
...
+import org.antlr.v4.runtime.CharStream;
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.TokenStream;
+import org.antlr.v4.runtime.*;
+import org.antlr.v4.runtime.atn.*;
+import org.antlr.v4.runtime.dfa.DFA;
+import org.antlr.v4.runtime.misc.*;
+
+@SuppressWarnings({"all", "warnings", "unchecked", "unused", "cast", "CheckReturnValue", "this-escape"})
+public class OpenSearchSQLLexer extends Lexer {
```
Comment (RyanL1997) [inline src/plugins/data/public/antlr/opensearch_sql/grammar/.antlr/OpenSearchSQLLexer.java:12]: For my knowledge, why we need this java file?

File: package.json [hunk 3]
```diff
@@ -84,7 +84,8 @@
     "cypress:run-without-security": "env TZ=America/Los_Angeles NO_COLOR=1 cypress run --headless --env SECURITY_ENABLED=false",
     "cypress:run-with-security": "env TZ=America/Los_Angeles NO_COLOR=1 cypress run --headless --env SECURITY_ENABLED=true,openSearchUrl=https://localhost:9200,WAIT_FOR_LOADER_BUFFER_MS=500",
     "osd:ciGroup10": "echo \"dashboard_sanity_test_spec.js\"",
-    "osd:ciGroup11": "echo \"apps/vis_builder/*.js\""
+    "osd:ciGroup11": "echo \"apps/vis_builder/*.js\"",
+    "generate:opensearchsqlantlr": "./node_modules/antlr4ng-cli/index.js -Dlanguage=TypeScript -o ./src/plugins/data/public/antlr/opensearch_sql/.generated -visitor -no-listener -Xexact-output-dir ./src/plugins/data/public/antlr/opensearch_sql/grammar/OpenSearchSQLLexer.g4 ./src/plugins/data/public/antlr/opensearch_sql/grammar/OpenSearchSQLParser.g4"
```
Comment (RyanL1997) [inline package.json:88]: Are this manual or auto generated? If it is manual, we may need better mechanism in the future to embed this into CI

File: src/plugins/data/public/antlr/opensearch_sql/grammar/OpenSearchSQLParser.g4 [hunk 4]
```diff
@@ -0,0 +1,849 @@
...
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+parser grammar OpenSearchSQLParser;
```
Comment (RyanL1997) [inline src/plugins/data/public/antlr/opensearch_sql/grammar/OpenSearchSQLParser.g4:31]: where is the source of this grammar? Is this coming from the sql backend?

**Review by RyanL1997 — APPROVED** [review]

File: src/plugins/data/public/antlr/shared/utils.ts (hunk 1 above)
Comment (mengweieric) [reply src/plugins/data/public/antlr/shared/utils.ts:24]: yes, and if it's undefined, the request will hit local cluster.

File: src/plugins/data/public/antlr/opensearch_sql/grammar/.antlr/OpenSearchSQLLexer.java (hunk 2 above)
Comment (mengweieric) [reply src/plugins/data/public/antlr/opensearch_sql/grammar/.antlr/OpenSearchSQLLexer.java:12]: these are auto-generated files from ANTLR engine for query parsing internally.

File: package.json (hunk 3 above)
Comment (mengweieric) [reply package.json:88]: for now it's manually generated, previously we add this to post commit hook but since the build pipeline does not have java runtime, we switch to manual and only need to run this when there's grammar changes

File: src/plugins/data/public/antlr/opensearch_sql/grammar/OpenSearchSQLParser.g4 (hunk 4 above)
Comment (mengweieric) [reply src/plugins/data/public/antlr/opensearch_sql/grammar/OpenSearchSQLParser.g4:31]: yes, it's from SQL plugin.

**Review by BionIT — APPROVED** [review]

File: src/plugins/data/public/antlr/shared/utils.ts (hunk 1 above)
Comment (RyanL1997) [reply src/plugins/data/public/antlr/shared/utils.ts:24]: Got it. Just wanna make sure the underlying logic can be relied on as a source of truth for the flag of `data_source.enabled`, since my work changed the registration behavior of datasource management plugin. And I've seen that some of the plugin used the datasource management plugin status to determine if multi-datasource is enabled or not.

**Review by RyanL1997 — APPROVED** [review]

**Comment by zhongnansu** [comment]
> need to unblock @paulstn for DQL, will follow up on increasing test coverage.

@mengweieric could you create a issue for tracking purpose?

**Review by zhongnansu — APPROVED** [review]

**Comment by mengweieric** [comment]
> > need to unblock @paulstn for DQL, will follow up on increasing test coverage.
> 
> @mengweieric could you create a issue for tracking purpose?

sure. https://github.com/opensearch-project/OpenSearch-Dashboards/issues/7387

File: src/plugins/data/public/ui/filter_bar/_global_filter_group.scss [hunk 5]
```diff
@@ -1,6 +1,7 @@
 // SASSTODO: Probably not the right file for this selector, but temporary until the files get re-organized
 .globalQueryBar {
   padding: 0 $euiSizeS $euiSizeS $euiSizeS;
+  height: 160px;
```
Comment (abbyhu2000) [inline src/plugins/data/public/ui/filter_bar/_global_filter_group.scss:4]: This line should not be added. This makes dashboard and visualize top nav bar being displayed incorrectly. 
Screenshot 2024-07-23 at 12 55 25 PM
Screenshot 2024-07-23 at 12 55 15 PM

File: src/plugins/data/public/ui/query_editor/language_switcher.tsx
Comment (abbyhu2000) [inline src/plugins/data/public/ui/query_editor/language_switcher.tsx:1]: Is this file needed? This is not used anywhere

File: src/plugins/data/public/ui/query_editor/_language_switcher.scss
Comment (abbyhu2000) [inline src/plugins/data/public/ui/query_editor/_language_switcher.scss:1]: And this too, it is not used anywhere

**Comment by AMoo-Miki** [comment]
This was manually backported to 2.x with https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7477 and to 2.16 with https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7487.
